dali/pipeline/data/buffer.h

```diff
@@ -289,34 +339,65 @@ class DLL_PUBLIC Buffer {
     set_type(TypeTable::GetTypeId<T>());
   }
 
-  inline void reserve(size_t new_num_bytes) {
-    if (new_num_bytes <= num_bytes_) return;
-
-    // re-allocating: get the device
-    if (std::is_same<Backend, GPUBackend>::value) {
-      CUDA_CALL(cudaGetDevice(&device_));
-    } else {
-      device_ = CPU_ONLY_DEVICE_ID;
-    }
+  inline void reserve(size_t new_num_bytes, AccessOrder order = {}) {
+    if (new_num_bytes <= num_bytes_)
+      return;
 
     DALI_ENFORCE(!shares_data_,
                  "Cannot reallocate Buffer if it is sharing data. "
                  "Clear the status by `Reset()` first.");
-    data_.reset();
+
+    if (data_ && device_ >= 0 && order.is_device() && order.device_id() != device_)
+      DALI_FAIL("Cannot reallocate a buffer on a different device!");
+
+    set_order(order);
+    free_storage();
+
+    if (device_ < 0) {
+      if (order.is_device() && order.device_id() >= 0) {
+        device_ = order.device_id();
+      } else {
+        // re-allocating: get the device
+        if (std::is_same<Backend, GPUBackend>::value || pinned_) {
+          device_ = order_.device_id();
+          if (device_ < 0)
+            CUDA_CALL(cudaGetDevice(&device_));
+        } else {
+          device_ = CPU_ONLY_DEVICE_ID;
+        }
+      }
+    }
+
     data_ = allocate_ ? allocate_(new_num_bytes)
-                      : AllocBuffer<Backend>(new_num_bytes, pinned_);
+                      : AllocBuffer<Backend>(new_num_bytes, pinned_, device_, order_);
 
     num_bytes_ = new_num_bytes;
   }
 
-  void reset() {
+  /**
+   * @brief Deallocates the data and clears the type.
+   *
+   * The data, if any, is deallocated. The function will
+   */
+  void reset(AccessOrder order = {}) {
```

Comment:
Shouldn't it use `order_`?

Reply:
If `order` is empty, it will: `free_storage` will use `order_` and `set_order` will be a no-op.

Reply:
Please add this in the docstring.